### ISSUE TITLE
Throttle refresh frequency when the API quota runs low

### ISSUE THREAD

**Issue by damithc** [HubTurbo/HubTurbo#1365]
Currently we refresh every 1 minute. We should lower the frequency if the API quota runs low.

**Comment by yl-coder** [comment]
how much should we increase when the api quota runs low?
>4000 : 1 mins
3000 - 4000 : 2 mins
2000 - 3000 : 3 mins
1000 - 2000 : 4 mins
0 - 1000 : 5mins 
?

**Comment by damithc** [comment]
It needs to factor in the remaining time, not just the remaining calls.

**Comment by yl-coder** [comment]
if apiQuota < 4000
  refreshTime = (5000 - apiQuota)/1000 * ONE_MINUTE (Similar to the one above)
else
 refreshTime = ONE_MINUTE

if refreshTime > remainingTime
   refreshTime = remainingTime

is this ok?

**Comment by damithc** [comment]
@acruis your views?

**Comment by acruis** [comment]
I disagree with fixed speeds based on api quota.

The api quota can be consumed at faster rates than the proposed tiers. Currently, it's possible to consume about 200 API calls per refresh, which can further increase.
At the same time, the user may also choose to modify his filters to adapt to the low api quota, in which case we should be able to return to the minutely refresh rate.

In other words, we should be able to dynamically modify the refresh rate; for example, based on the amount of quota consumed by the most recent refresh.

**Comment by damithc** [comment]
Yes, it needs to be dynamic. For example, if the most recent refresh took 200 calls and there are only 2000 calls left for the next 40 minutes, we have to refresh every 4 minutes.

**Comment by yl-coder** [comment]
got it.

Ok got it. I will use this. i guess we need to set a min boundary? Highest freq will be 1 min.

refreshTime = timeRemaining/(apiQuota/prevAPICallsConsumed)

//Set min boundary
if(refreshTime < 60)
   refreshTime = ONE_MINUTE

**Comment by damithc** [comment]
Adjust the algorithm such that we aim to save about 200 api calls at the end. This will serve as a buffer to be used for write operations and/or manual refresh. Perhaps we can also show the current interval somewhere, may be the same place we show the quota status?

**Comment by yl-coder** [comment]
ok. Also for the quota status, I think we should change how we display the quota and time remaining. Currently it is not very intuitive. Maybe we should include descriptive text? Or even better, to show it graphically. Like using a progress bar to indicate those details.

Views?

**Comment by damithc** [comment]
It can remain cryptic for now. It's for our own monitoring and currently in a temp location. We can make it more intuitive at a later stage, once we move it to a more permanent location. So something like `2213/34[x2]` is fine (x2 means every 2 minutes)

**Comment by yl-coder** [comment]
got it. will implement it
